### PR TITLE
 refactor(feedback-api): duplicate adminform feedback endpoints for /api/v3

### LINKED ISSUE #1508: [API] Duplicate admin's form feedback related endpoints to their new /api counterparts
This issue tracks the straight duplication of admin form feedback related endpoints to their `/api` rooted endpoints to follow a RESTful hierarchy. 

The old endpoint should not be deleted for backwards compatibility (at least for a few weeks after the new endpoint has been released).

The current to new mapping is shown below:

\# | Existing endpoints | New endpoints<br>(note that all routes are prefixed with /api/v3) | Description
-- | -- | --|--
1 | GET /:formId/adminform/feedback | GET /admin/forms/:formId/feedback | retrieve all form feedback
2 | GET /:formId/adminform/feedback/count | GET /admin/forms/:formId/feedback/count | get feedback count
3 | GET /:formId/adminform/feedback/download | GET /admin/forms/:formId/feedback/download | download form feedback

### PR DESCRIPTION
**Pull request by orbitalsqwib**
## Problem

Update current adminform feedback endpoints to their /api rooted endpoints to follow a RESTful hierarchy.

Closes #1508 

## Solution

**Features**:

- shard admin routes into separate files
- duplicate adminform feedback endpoint functionalities and integration tests with the new api v3 routes
- update frontend api calls to use new endpoints

## Tests

- [ ] Create a new form, submit a new response, then submit feedback for the form
- [ ] Return to the form admin feedback section
- [ ] Check that the feedback can be successfully retrieved (GET /api/v3/admin/forms/:formId/feedback returns 200)
- [ ] Check that the form feedback count is correct (GET /api/v3/admin/forms/:formId/feedback/count returns 200)
- [ ] Check that the feedback export function works (GET /api/v3/admin/forms/:formId/feedback/download returns 200)
- [ ] Check that calling GET /api/v3/admin/forms/:formId/feedback without logging in results in a 401 Unauthorized error
- [ ] Check that calling PATCH /api/v3/admin/forms/:formId/settings without logging in results in a 401 Unauthorized error

### CODE REVIEW

File: src/app/routes/api/v3/admin/forms/__tests__/admin-forms.feedback.routes.spec.ts [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+import { AdminFormsFeedbackRouter } from '../admin-forms.feedback.routes'
+
+// Prevent rate limiting.
+jest.mock('src/app/utils/limit-rate')
+
+const UserModel = getUserModel(mongoose)
+const EncryptFormModel = getEncryptedFormModel(mongoose)
+const FormFeedbackModel = getFormFeedbackModel(mongoose)
+
+const app = setupApp(undefined, AdminFormsFeedbackRouter, {
```
Comment (karrui) [inline src/app/routes/api/v3/admin/forms/__tests__/admin-forms.feedback.routes.spec.ts:29]: can we have this router be set up with its real root route like `/admin/forms`?

File: src/app/routes/api/v3/admin/forms/__tests__/admin-forms.feedback.routes.spec.ts [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+      formForFeedback = (await EncryptFormModel.create({
+        title: 'form to view feedback',
+        admin: defaultUser._id,
+        publicKey: 'does not matter',
+      })) as IFormDocument
+    })
+
+    it('should return 200 with empty feedback meta when no feedback exists', async () => {
+      // Act
+      const response = await request.get(`/${formForFeedback._id}/feedback`)
```
Comment (karrui) [inline src/app/routes/api/v3/admin/forms/__tests__/admin-forms.feedback.routes.spec.ts:63]: then here we test the request with `/admin/forms/:formId/feedback` so we see the full route

File: src/app/routes/api/v3/admin/forms/admin-forms.feedback.routes.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import { Router } from 'express'
+
+import { withUserAuthentication } from '../../../../../modules/auth/auth.middlewares'
+import * as AdminFormController from '../../../../../modules/form/admin-form/admin-form.controller'
+
+export const AdminFormsFeedbackRouter = Router()
+
```
Comment (karrui) [inline src/app/routes/api/v3/admin/forms/admin-forms.feedback.routes.ts:7]: you're leaving this router unprotected by auth middleware

File: src/app/routes/api/v3/admin/forms/admin-forms.settings.routes.ts [hunk 4]
```diff
@@ -6,10 +6,10 @@ import { SettingsUpdateDto } from '../../../../../../types/api'
 import { withUserAuthentication } from '../../../../../modules/auth/auth.middlewares'
 import { handleUpdateSettings } from '../../../../../modules/form/admin-form/admin-form.controller'
 
-export const AdminFormsRouter = Router()
+export const AdminFormsSettingsRouter = Router()
 
 // All routes in this handler should be protected by authentication.
-AdminFormsRouter.use(withUserAuthentication)
+AdminFormsSettingsRouter.use(withUserAuthentication)
```
Comment (karrui) [inline src/app/routes/api/v3/admin/forms/admin-forms.settings.routes.ts:12]: that's the danger with splitting up the routes like this during refactors haha.

i suggest putting the `withUserAuthentication` middleware into the index router

File: src/app/routes/api/v3/admin/forms/index.ts [hunk 5]
```diff
@@ -1 +1,9 @@
-export { AdminFormsRouter } from './admin-forms.routes'
+import { Router } from 'express'
+
+import { AdminFormsFeedbackRouter } from './admin-forms.feedback.routes'
+import { AdminFormsSettingsRouter } from './admin-forms.settings.routes'
+
+export const AdminFormsRouter = Router()
```
Comment (karrui) [inline src/app/routes/api/v3/admin/forms/index.ts:6]: use the `withUserAuthentication` middleware here, or even in the root `/admin` router.

**Review by karrui — CHANGES_REQUESTED** [review]

File: src/app/routes/api/v3/admin/forms/__tests__/admin-forms.feedback.routes.spec.ts (hunk 1 above)
Comment (orbitalsqwib) [reply src/app/routes/api/v3/admin/forms/__tests__/admin-forms.feedback.routes.spec.ts:29]: sure, will do

File: src/app/routes/api/v3/admin/forms/admin-forms.feedback.routes.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+ * @returns 200 with feedback response
+ * @returns 401 when user does not exist in session
+ * @returns 403 when user does not have permissions to access form
+ * @returns 404 when form cannot be found
+ * @returns 410 when form is archived
+ * @returns 422 when user in session cannot be retrieved from the database
+ * @returns 500 when database error occurs
+ */
+AdminFormsFeedbackRouter.get(
+  '/:formId([a-fA-F0-9]{24})/feedback',
```
Comment (karrui) [inline src/app/routes/api/v3/admin/forms/admin-forms.feedback.routes.ts:21]: so sorry, only spotted this when reviewing; can we change this (and all other) endpoints from `/feedback` to `/feedbacks` (plural for consistency)? sorry for the multiple changes due to my mistake in writing the issue

Comment (karrui) [reply src/app/routes/api/v3/admin/forms/admin-forms.feedback.routes.ts:21]: I have updated the original issue.

Comment (orbitalsqwib) [reply src/app/routes/api/v3/admin/forms/admin-forms.feedback.routes.ts:21]: np

Comment (mantariksh) [reply src/app/routes/api/v3/admin/forms/admin-forms.feedback.routes.ts]: wait `feedbacks` is not a word though

**Comment by karrui** [comment]
Feedback is an uncountable noun, sorry for the fail

File: src/app/routes/api/v3/admin/forms/admin-forms.feedback.routes.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  AdminFormController.handleGetFormFeedbacks,
```
Comment (karrui) [inline src/app/routes/api/v3/admin/forms/admin-forms.feedback.routes.ts:22]: nit: should we rename this to `handleGetFormFeedback` instead too?

**Review by karrui — APPROVED** [review]
lgtm

File: src/app/routes/api/v3/admin/forms/admin-forms.feedback.routes.ts (hunk 7 above)
Comment (orbitalsqwib) [reply src/app/routes/api/v3/admin/forms/admin-forms.feedback.routes.ts:22]: sure

**Review by karrui — APPROVED** [review]

**Pull request merged**